content/2016.12.how-to-develop-a-webapp-with-Kinto.md

```diff
@@ -0,0 +1,256 @@
+* We will need a bit of JavaScript to enhance and animate interactions within the application and we will also need a way to store and share the app's data. Ideally, we would like our application to consume a REST API exposing endpoints like `https://api.discute.re/v1/rooms/12345/messages`, where `discute.re` is the domain of our application (*discutere* means to chat in Latin).
+
+Our API endpoints would be:
+
+* `/rooms` to return all chat rooms,
+* `/rooms/:id/` to return a particular chat room identified by an ID,
+* `/rooms/:id/messages` to return all messages from a particular room.
+
+To post a message, our front-end will do a POST request to `/rooms/:id/messages`.
+
+We would store each message like this:
+
+```json
+{
+    "message": "I know nothing",
+    "last_modified": 1480340246147,
+    "author": "Jon Snow",
+    "id": "a1e6e879-5c94-4a01-9ec2-28e152d1d8f4"
+}
+```
+
+The JavaScript code could use whatever library or latest hype JS framework but in our example we will stick with modern [Vanilla JS](http://vanilla-js.com/) and use top notch [ES2015 syntax](https://babeljs.io/docs/learn-es2015/).
+
+## API creation
+
+Hmm… what backend should we use? Django? Express? Firebase?… oh wait, why don’t we give [Kinto](http://www.kinto-storage.org/) a try?
+
+Kinto provides a flexible and generic enough API to address the needs of many different kinds of application. Call us lazy, effective or pragmatic but Kinto will prevent us to write our own API and help us handle other APIs.
+
+Let's see how to use Kinto for our instant chat WebApp.
+
+To show some pride, we will use an explicit subdomain for our main URL `https://kinto.discute.re/v1` to show our configuration rather than an agnostic `https://api.discute.re/v1`.
+
+Using Kinto, here are the native API endpoints we will use to match our application requirements:
+
+- `/rooms/4815162342` becomes `/collections/4815162342` ;
+- `/rooms/4815162342/messages` becomes `/collections/4815162342/records` ;
+
+As we have seen before, we can add a new post to our backend with a simple POST request
+`/collections/4815162342/records`.
+
+A Kinto server can handle many Web applications through the concept of [*buckets*](https://kinto.readthedocs.io/en/latest/concepts.html#buckets-collections-and-records); a bucket can contain many *collections*, a collection can contain many *records*, and records can contain arbitrary JSON data.
+
+![](https://kinto.readthedocs.io/en/latest/_images/concepts-general.png)
+{.image}
+
+For our application, we’ll create a *chat* bucket, which will contain as many collections as we want distinct *chat rooms*. Each room collection will itself represent *chat messages* through records.
+
+You can use the [Kinto Admin](http://kinto.github.io/kinto-admin/) to visually browse and manage your application data.
+
+![Our records in Kinto Admin](images/kinto-admin-webchat-example.png)
+{.image}
+
+Considering a chat room with the id **4815162342** (*Lost numbers*), we could:
```

Comment:
I think this is joke could stand without subtracting from focus if it were an "Easter egg", for example `[4815162342](https://en.wikipedia.org/wiki/Mythology_of_Lost#The_Numbers)`. I never watched Lost but personally I think links to only tangentially related pages add to the richness of the Web.